File: demo-python/interactive_agent.py
#!/usr/bin/env python3
"""
交互式AI Agent界面
"""
import os
import sys
import time
import signal
from typing import Dict, Any
from colorama import init, Fore, Style, Back
from agent import chat_with_agent, config, conversation_history

# 初始化colorama
init(autoreset=True)

# 会话状态
current_thread_id = "1"
is_running = True

def signal_handler(signum, frame):
    """信号处理器"""
    global is_running
    print(f"\n{Fore.YELLOW}正在退出...{Style.RESET_ALL}")
    is_running = False
    sys.exit(0)

# 注册信号处理器
signal.signal(signal.SIGINT, signal_handler)

def print_separator(char="━", length=70, color=Fore.CYAN):
    """打印分隔线"""
    print(f"{color}{char * length}{Style.RESET_ALL}")

def print_box(content: str, title: str = "", color=Fore.CYAN):
    """打印边框"""
    lines = content.strip().split('\n')
    max_length = max(len(line) for line in lines)
    
    # 标题
    if title:
        print(f"{color}╔{'═' * (max_length + 2)}╗{Style.RESET_ALL}")
        print(f"{color}║{title.center(max_length + 2)}║{Style.RESET_ALL}")
        print(f"{color}╠{'═' * (max_length + 2)}╣{Style.RESET_ALL}")
    else:
        print(f"{color}╔{'═' * (max_length + 2)}╗{Style.RESET_ALL}")
    
    # 内容
    for line in lines:
        print(f"{color}║{Style.RESET_ALL} {line:<{max_length}} {color}║{Style.RESET_ALL}")
    
    # 底部
    print(f"{color}╚{'═' * (max_length + 2)}╝{Style.RESET_ALL}")

def show_welcome():
    """显示欢迎界面"""
    os.system('clear' if os.name == 'posix' else 'cls')
    
    welcome_text = """
🤖 增强型 AI Agent 交互式界面

🚀 功能特性:
• 🔍 网络搜索 (TavilySearch)
• 🧮 数学计算 (Calculator)
• ⏰ 时间查询
• 📁 文件操作 (读取/写入/列表)
• 🌤️ 天气查询
• 💻 系统信息
• 🧠 对话记忆

🔧 命令:
• /help     - 显示帮助
• /clear    - 清除屏幕
• /config   - 显示配置
• /history  - 显示对话历史
• /reset    - 重置对话历史
• /exit     - 退出程序

💡 提示: 直接输入问题开始对话
"""
    
    print_box(welcome_text, "🤖 AI Agent", Fore.BLUE)

def show_help():
    """显示帮助信息"""
    help_text = """
🔧 命令说明:
  /help     - 显示此帮助信息
  /clear    - 清除屏幕
  /config   - 显示当前配置
  /history  - 显示对话历史
  /reset    - 重置对话历史
  /exit     - 退出程序

📝 使用示例:
  • 现在几点了？
  • 帮我计算 123 + 456 * 789
  • 搜索最新的AI技术发展
  • 北京今天天气怎么样？
  • 读取README.md文件
  • 获取系统信息
"""
    
    print_box(help_text, "📖 帮助", Fore.GREEN)

def show_config():
    """显示配置信息"""
    config_text = f"""
模型: {config['llm']['model']}
温度: {config['llm']['temperature']}
最大令牌: {config['llm']['max_tokens']}
搜索结果数: {config['search']['max_results']}
当前线程: {current_thread_id}

启用的工具:
"""
    
    enabled_tools = [key for key, value in config["tools"].items() if value.get("enabled", False)]
    config_text += "  • " + "\n  • ".join(enabled_tools)
    
    print_box(config_text, "🔧 当前配置", Fore.MAGENTA)

def show_history():
    """显示对话历史"""
    if not conversation_history:
        print(f"{Fore.YELLOW}📝 暂无对话历史{Style.RESET_ALL}")
        return
    
    print(f"{Fore.CYAN}📚 对话历史:{Style.RESET_ALL}")
    print_separator()
    
    for i, msg in enumerate(conversation_history[-10:]):  # 显示最近10条
        role = msg["role"]
        content = msg["content"]
        
        if len(content) > 100:
            content = content[:100] + "..."
        
        role_color = Fore.BLUE if role == "用户" else Fore.GREEN
        print(f"{role_color}[{role}]:{Style.RESET_ALL} {content}")
        
        if i < len(conversation_history[-10:]) - 1:
            print(f"{Fore.CYAN}{'─' * 50}{Style.RESET_ALL}")
    
    print_separator()

def reset_history():
    """重置对话历史"""
    global conversation_history
    conversation_history.clear()
    print(f"{Fore.GREEN}✅ 对话历史已重置{Style.RESET_ALL}")

def process_command(command: str) -> bool:
    """处理命令"""
    cmd_parts = command[1:].split()
    cmd = cmd_parts[0] if cmd_parts else ""
    
    if cmd == "help":
        show_help()
    elif cmd == "clear":
        show_welcome()
    elif cmd == "config":
        show_config()
    elif cmd == "history":
        show_history()
    elif cmd == "reset":
        reset_history()
    elif cmd == "exit":
        print(f"{Fore.GREEN}👋 再见！{Style.RESET_ALL}")
        return False
    else:
        print(f"{Fore.RED}❌ 未知命令: {cmd}。输入 /help 查看可用命令。{Style.RESET_ALL}")
    
    return True

def show_thinking():
    """显示思考动画"""
    thinking_chars = ["🤔", "💭", "🧠", "⚡"]
    for i in range(4):
        print(f"\r{thinking_chars[i % len(thinking_chars)]} 思考中...", end="", flush=True)
        time.sleep(0.5)
    print("\r" + " " * 20, end="\r")  # 清除思考提示

def main():
    """主函数"""
    global is_running
    
    # 显示欢迎信息
    show_welcome()
    
    while is_running:
        try:
            # 获取用户输入
            user_input = input(f"\n{Fore.CYAN}🤖 > {Style.RESET_ALL}").strip()
            
            if not user_input:
                continue
            
            # 处理命令
            if user_input.startswith("/"):
                if not process_command(user_input):
                    break
                continue
            
            # 处理普通对话
            print(f"\n{Fore.BLUE}用户:{Style.RESET_ALL} {user_input}")
            print_separator("─", 50, Fore.CYAN)
            
            # 显示思考动画
            show_thinking()
            
            # 获取响应
            response = chat_with_agent(user_input)
            
            # 显示响应
            print(f"{Fore.GREEN}🤖 Agent:{Style.RESET_ALL}")
            print(response)
            
            # 显示时间戳
            if config["ui"]["show_timestamp"]:
                timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
                print(f"{Style.DIM}⏱️ {timestamp}{Style.RESET_ALL}")
            
            print_separator("─", 50, Fore.CYAN)
            
        except KeyboardInterrupt:
            print(f"\n{Fore.YELLOW}正在退出...{Style.RESET_ALL}")
            break
        except EOFError:
            print(f"\n{Fore.YELLOW}程序已退出{Style.RESET_ALL}")
            break
        except Exception as e:
            print(f"{Fore.RED}❌ 错误: {e}{Style.RESET_ALL}")
    
    print(f"\n{Fore.GREEN}程序已退出{Style.RESET_ALL}")

if __name__ == "__main__":
    main() 

File: demo-js/interactive-agent.js
import 'dotenv/config'
import readline from 'readline'
import chalk from 'chalk'
import { TavilySearch } from '@langchain/tavily'
import { ChatAnthropic } from '@langchain/anthropic'
import { MemorySaver } from '@langchain/langgraph'
import { createReactAgent } from '@langchain/langgraph/prebuilt'
import { Calculator } from '@langchain/community/tools/calculator'
import { DynamicTool } from '@langchain/core/tools'
import { HumanMessage } from '@langchain/core/messages'
import { getConfig, validateConfig } from './agent-config.js'
import fs from 'fs'

// 获取配置
const config = getConfig()

// 验证配置
const configErrors = validateConfig(config)
if (configErrors.length > 0) {
  console.error(chalk.red('配置错误:'))
  configErrors.forEach(error => console.error(chalk.red(`  - ${error}`)))
  process.exit(1)
}

// 创建自定义工具
const getCurrentTime = new DynamicTool({
  name: 'getCurrentTime',
  description: '获取当前日期和时间',
  func: async () => {
    const now = new Date()
    return `当前时间: ${now.toLocaleString('zh-CN', { 
      timeZone: config.tools.time.timezone,
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit'
    })}`
  }
})

const fileOperations = new DynamicTool({
  name: 'fileOperations',
  description: '读取、写入或列出文件。输入格式：{"action": "read|write|list", "path": "文件路径", "content": "写入内容(仅写入时需要)"}',
  func: async (input) => {
    try {
      const { action, path: filePath, content } = JSON.parse(input)
      
      // 安全检查
      if (config.security.blockedDirectories.some(dir => filePath.includes(dir))) {
        return '访问被禁止的目录'
      }
      
      // 检查文件大小
      if (action === 'read' && fs.existsSync(filePath)) {
        const stats = fs.statSync(filePath)
        if (stats.size > config.fileOperations.maxFileSize) {
          return '文件过大，无法读取'
        }
      }
      
      switch (action) {
        case 'read':
          if (fs.existsSync(filePath)) {
            return fs.readFileSync(filePath, config.fileOperations.defaultEncoding)
          } else {
            return '文件不存在'
          }
        
        case 'write':
          // 备份现有文件
          if (config.fileOperations.backupEnabled && fs.existsSync(filePath)) {
            fs.copyFileSync(filePath, `${filePath}.backup`)
          }
          fs.writeFileSync(filePath, content, config.fileOperations.defaultEncoding)
          return `文件已写入: ${filePath}`
        
        case 'list':
          if (fs.existsSync(filePath)) {
            const files = fs.readdirSync(filePath)
            return `目录内容: ${files.join(', ')}`
          } else {
            return '目录不存在'
          }
        
        default:
          return '不支持的操作，请使用 read、write 或 list'
      }
    } catch (error) {
      return `操作失败: ${error.message}`
    }
  }
})

const weatherTool = new DynamicTool({
  name: 'weatherTool',
  description: '获取指定城市的详细天气信息',
  func: async (cityName) => {
    const mockWeatherData = {
      '北京': { temperature: '15°C', humidity: '65%', windSpeed: '10km/h', condition: '晴' },
      '上海': { temperature: '18°C', humidity: '70%', windSpeed: '8km/h', condition: '多云' },
      '广州': { temperature: '22°C', humidity: '80%', windSpeed: '12km/h', condition: '阴' },
      '深圳': { temperature: '24°C', humidity: '85%', windSpeed: '15km/h', condition: '雨' }
    }
    
    const weather = mockWeatherData[cityName] || { 
      temperature: '未知', 
      humidity: '未知', 
      windSpeed: '未知', 
      condition: '数据不可用' 
    }
    
    return `${cityName}天气：
温度: ${weather.temperature}
湿度: ${weather.humidity}
风速: ${weather.windSpeed}
天气状况: ${weather.condition}`
  }
})

const systemInfo = new DynamicTool({
  name: 'systemInfo',
  description: '获取系统信息',
  func: async () => {
    const os = await import('os')
    const info = `系统信息:
操作系统: ${os.platform()} ${os.release()}
架构: ${os.arch()}
总内存: ${(os.totalmem() / 1024 / 1024 / 1024).toFixed(2)} GB
可用内存: ${(os.freemem() / 1024 / 1024 / 1024).toFixed(2)} GB
Node.js版本: ${process.version}
运行时间: ${Math.floor(process.uptime())} 秒`
    
    // 过滤敏感信息
    if (config.tools.system.sensitiveInfoFilter) {
      return info.replace(/\/Users\/[^\/]+/g, '/Users/***')
    }
    
    return info
  }
})

// 创建工具数组
const agentTools = [
  ...(config.tools.search.enabled ? [new TavilySearch({
    maxResults: config.search.maxResults,
    searchDepth: config.search.searchDepth
  })] : []),
  ...(config.tools.calculator.enabled ? [new Calculator()] : []),
  ...(config.tools.time.enabled ? [getCurrentTime] : []),
  ...(config.tools.fileOps.enabled ? [fileOperations] : []),
  ...(config.tools.weather.enabled ? [weatherTool] : []),
  ...(config.tools.system.enabled ? [systemInfo] : [])
]

// 创建LLM
const agentModel = new ChatAnthropic({
  model: config.llm.model,
  temperature: config.llm.temperature,
  maxTokens: config.llm.maxTokens
})

// 创建记忆
const agentCheckpoint = new MemorySaver()

// 创建Agent
const agent = createReactAgent({
  llm: agentModel,
  tools: agentTools,
  checkpointSaver: agentCheckpoint,
})

// 创建命令行界面
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  prompt: chalk.cyan('🤖 > ')
})

// 会话状态
let currentThreadId = '1'
let isRunning = true

// 日志记录
function log(level, message) {
  if (config.logging.enableConsole) {
    const timestamp = new Date().toISOString()
    console.log(`[${timestamp}] [${level.toUpperCase()}] ${message}`)
  }
}

// 显示启动信息
function showWelcome() {
  console.log(chalk.cyan.bold(`
╔════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║                                                                                                                        ║
║                                  🤖 增强型 AI Agent 交互式界面                                                       ║
║                                                                                                                        ║
║  🚀 功能特性:                                                                                                          ║
║  • 🔍 网络搜索 (TavilySearch)                                                                                        ║
║  • 🧮 数学计算 (Calculator)                                                                                          ║
║  • ⏰ 时间查询                                                                                                        ║
║  • 📁 文件操作 (读取/写入/列表)                                                                                      ║
║  • 🌤️ 天气查询                                                                                                       ║
║  • 💻 系统信息                                                                                                       ║
║  • 🧠 对话记忆                                                                                                       ║
║                                                                                                                        ║
║  🔧 命令:                                                                                                              ║
║  • /help     - 显示帮助                                                                                              ║
║  • /clear    - 清除屏幕                                                                                              ║
║  • /config   - 显示配置                                                                                              ║
║  • /thread   - 切换对话线程                                                                                          ║
║  • /history  - 显示对话历史                                                                                          ║
║  • /exit     - 退出程序                                                                                              ║
║                                                                                                                        ║
╚════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝
  `))
}

// 显示帮助
function showHelp() {
  console.log(chalk.yellow(`
🔧 命令说明:
  /help     - 显示此帮助信息
  /clear    - 清除屏幕
  /config   - 显示当前配置
  /thread <id> - 切换到指定对话线程
  /history  - 显示对话历史
  /exit     - 退出程序

📝 使用示例:
  • 现在几点了？
  • 帮我计算 123 + 456 * 789
  • 搜索最新的AI技术发展
  • 北京今天天气怎么样？
  • 读取README.md文件
  • 获取系统信息
  `))
}

// 显示配置
function showConfig() {
  console.log(chalk.magenta(`
🔧 当前配置:
  • 模型: ${config.llm.model}
  • 温度: ${config.llm.temperature}
  • 最大令牌: ${config.llm.maxTokens}
  • 搜索结果数: ${config.search.maxResults}
  • 记忆文件: ${config.memory.memoryFile}
  • 当前线程: ${currentThreadId}
  • 启用的工具: ${Object.keys(config.tools).filter(key => config.tools[key].enabled).join(', ')}
  `))
}

// 处理用户输入
async function handleInput(input) {
  const command = input.trim()
  
  // 处理命令
  if (command.startsWith('/')) {
    const [cmd, ...args] = command.slice(1).split(' ')
    
    switch (cmd) {
      case 'help':
        showHelp()
        break
      
      case 'clear':
        console.clear()
        showWelcome()
        break
      
      case 'config':
        showConfig()
        break
      
      case 'thread':
        if (args[0]) {
          currentThreadId = args[0]
          console.log(chalk.green(`已切换到线程: ${currentThreadId}`))
        } else {
          console.log(chalk.yellow(`当前线程: ${currentThreadId}`))
        }
        break
      
      case 'history':
        console.log(chalk.blue('对话历史功能开发中...'))
        break
      
      case 'exit':
        console.log(chalk.green('再见！👋'))
        isRunning = false
        rl.close()
        return
      
      default:
        console.log(chalk.red(`未知命令: ${cmd}。输入 /help 查看可用命令。`))
    }
    
    rl.prompt()
    return
  }
  
  // 处理正常对话
  if (command) {
    try {
      console.log(chalk.gray('🤔 思考中...'))
      
      const startTime = Date.now()
      const result = await agent.invoke(
        { messages: [new HumanMessage(command)] },
        { configurable: { thread_id: currentThreadId } }
      )
      const endTime = Date.now()
      
      const response = result.messages[result.messages.length - 1].content
      
      console.log(chalk.green(`🤖 Agent: ${response}`))
      
      if (config.ui.showTimestamp) {
        console.log(chalk.gray(`⏱️ 响应时间: ${endTime - startTime}ms`))
      }
      
      log('info', `User: ${command}`)
      log('info', `Agent: ${response}`)
      
    } catch (error) {
      console.error(chalk.red(`❌ 错误: ${error.message}`))
      log('error', `Error: ${error.message}`)
    }
  }
  
  rl.prompt()
}

// 启动交互式界面
function startInteractiveAgent() {
  console.clear()
  showWelcome()
  
  rl.on('line', handleInput)
  
  rl.on('close', () => {
    console.log(chalk.yellow('\n程序已退出'))
    process.exit(0)
  })
  
  rl.prompt()
}

// 信号处理
process.on('SIGINT', () => {
  console.log(chalk.yellow('\n正在退出...'))
  rl.close()
})

// 启动程序
startInteractiveAgent() 